README.md
cors-gate
=========

[![Build Status](https://travis-ci.org/mixmaxhq/cors-gate.svg?branch=master)](https://travis-ci.org/mixmaxhq/cors-gate)

Connect-compatible middleware to selectively reject requests based on CORS rules.

This lets you implement an elegant alternative to CSRF tokens if you only need to support
modern browsers. For more information, see [our blog post](https://mixmax.com/blog/modern-csrf).

Install
-------

Run this in your project:

```sh
$ yarn add cors-gate
```

Test
----

```sh
$ yarn test
```

Usage
-----

```js
const express = require('express');
const cors = require('cors');
const corsGate = require('cors-gate');

const app = express();

app.use(cors({
  origin: ['https://app.mixmax.com', 'https://other-app.mixmax.com'],
  credentials: true
}));

// prevent cross-origin requests from domains not permitted by the preceeding cors rules
app.use(corsGate({
  // require an Origin header, and reject request if missing
  strict: true,
  // permit GET and HEAD requests, even without an Origin header
  allowSafe: true,
  // the origin of the server
  origin: 'https://api.mixmax.com'
}));

// add a new contact
app.post('/api/contacts', function(req, res) {
  // ...
  res.status(200).json({id: id});
});
```

### Alternative failure handling

By default, `cors-gate` will return `403 Unauthorized` to any requests that aren't permitted by the specified options.

The `failure` option offers a means to change this behavior. This way, unauthorized cross-origin requests can be permitted in a restricted manner - perhaps by requiring an explicit authentication mechanism rather than cookie-based authentication to prevent cross-site scripting. As such, `cors-gate` can serve as a CSRF mechanism without the need for a token, while still allowing limited forms of third-party cross-origin API requests.

```js
app.use(corsGate({
  origin: 'https://api.mixmax.com',
  failure: function(req, res, next) {
    // requests from other origins will have this flag set.
    req.requireExplicitAuthentication = true;
    next();
  }
}));
```

### Excluding routes

Optionally you can make some paths unprotected as follows:

```js
app.use(corsGate({ origin: 'https://api.mixmax.com', }).unless({path: ['/webhook']}));
```

### Firefox and the Origin header

Firefox does not set the `Origin` header [on same-origin requests](http://stackoverflow.com/a/15514049/495611) (see also [csrf-request-tester](https://github.com/mixmaxhq/csrf-request-tester)) for same-origin requests, as of version 53. The `corsGate.originFallbackToReferrer` middleware will, if the `Origin` header is missing, fill it with the origin part of the `Referer`. This middleware thus enables verification of the `Origin` for same-origin requests.

Additionally, no browser sends the `Origin` header when sending a `GET` request to load an image. We could simply allow all `GET` requests - `GET` requests are safe, per `HTTP` - but we'd rather reject unauthorized cross-origin `GET` requests wholesale.

At present, Chrome and Safari do not support the `strict-origin` `Referrer-Policy`, so we can only patch the `Origin` from the `Referer` on Firefox. In patching it, however, we can reject unauthorized cross-origin `GET` requests from images, and once Chrome and Safari support `strict-origin`, we'll be able to do so on all three platforms.

In order to actually reject these requests, however, the patched `Origin` data must be visible to the `cors` middleware. This middleware is distinct because it must appear before `cors` and `corsGate` to perform all the described tasks.

```js
app.use(corsGate.originFallbackToReferrer());
app.use(cors({ ... }));
app.use(corsGate({ ... }));
```

### Language ports

- Go: [Roemerb/corsgate](https://github.com/Roemerb/corsgate)

License
-------

The [MIT License](https://github.com/mixmaxhq/cors-gate/blob/master/LICENSE).


CHANGELOG.md
## Release History

* 2.0.2 Fix for calling success without returning

* 2.0.1 Re-publishing 2.0

* 2.0.0 Terminate OPTIONS requests

* 1.1.2 Terminate preflight requests

* 1.1.1 Terminate OPTION requests

* 1.1.0 Allow a function to be passed for allowSafe


spec/tests.js
const alice = 'http://localhost:9997',
  bob = 'http://localhost:9998',
  eve = 'http://localhost:9999';

describe('cors-gate', function() {
  describe('same-origin', function() {
    it('should allow image requests', function(done) {
      dispatch({
        frameSrc: alice + '/',
        type: 'image',
        url: '/api/image'
      }, done);
    });

    it('should allow GET requests', function(done) {
      dispatch({
        frameSrc: alice + '/',
        type: 'xhr',
        method: 'GET',
        url: '/api/get'
      }, done);
    });

    it('should allow POST requests', function(done) {
      dispatch({
        frameSrc: alice + '/',
        type: 'xhr',
        method: 'POST',
        url: '/api/post',
        data: 'a=b'
      }, done);
    });

    it('should allow JSON POST requests', function(done) {
      dispatch({
        frameSrc: alice + '/',
        type: 'xhr',
        method: 'POST',
        url: '/api/post',
        headers: {
          'content-type': 'application/json'
        },
        data: '{"a":"b"}'
      }, done);
    });

    if (hasBeacon())
    xit('should allow POST requests by beacon', function(done) {
      dispatch({
        frameSrc: alice + '/',
        type: 'beacon',
        url: '/api/post',
        data: '{"a":"b"}'
      }, done);
    });
  });

  describe('cross-origin', function() {
    it('should allow authorized image requests', function(done) {
      dispatch({
        frameSrc: bob + '/',
        type: 'image',
        url: alice + '/api/image'
      }, done);
    });

    it('should allow authorized GET requests', function(done) {
      dispatch({
        frameSrc: bob + '/',
        type: 'xhr',
        method: 'GET',
        url: alice + '/api/get'
      }, done);
    });

    it('should allow authorized POST requests', function(done) {
      dispatch({
        frameSrc: bob + '/',
        type: 'xhr',
        method: 'POST',
        url: alice + '/api/post',
        data: 'a=b'
      }, done);
    });

    it('should allow authorized JSON POST requests', function(done) {
      dispatch({
        frameSrc: bob + '/',
        type: 'xhr',
        method: 'POST',
        url: alice + '/api/post',
        headers: {
          'content-type': 'application/json'
        },
        data: '{"a":"b"}'
      }, done);
    });

    if (hasBeacon())
    xit('should allow authorized POST requests by beacon', function(done) {
      dispatch({
        frameSrc: bob + '/',
        type: 'beacon',
        url: alice + '/api/post',
        data: '{"a":"b"}'
      }, done);
    });

    /*it('should reject unauthorized image requests', function(done) {
      dispatch({
        frameSrc: eve + '/',
        type: 'image',
        url: alice + '/api/image'
      }, expectFailure(done));
    });

    it('should reject unauthorized GET requests', function(done) {
      dispatch({
        frameSrc: eve + '/',
        type: 'xhr',
        method: 'GET',
        url: alice + '/api/get'
      }, expectFailure(done));
    });*/

    it('should reject unauthorized POST requests', function(done) {
      dispatch({
        frameSrc: eve + '/',
        type: 'xhr',
        method: 'POST',
        url: alice + '/api/post',
        data: 'a=b'
      }, expectFailure(done));
    });

    it('should reject unauthorized JSON POST requests', function(done) {
      dispatch({
        frameSrc: eve + '/',
        type: 'xhr',
        method: 'POST',
        url: alice + '/api/post',
        headers: {
          'content-type': 'application/json'
        },
        data: '{"a":"b"}'
      }, expectFailure(done));
    });

    if (hasBeacon())
    xit('should reject unauthorized POST requests by beacon', function(done) {
      dispatch({
        frameSrc: eve + '/',
        type: 'beacon',
        url: alice + '/api/post',
        data: '{"a":"b"}'
      }, expectFailure(done));
    });
  });
});

function hasBeacon() {
  return typeof navigator.sendBeacon === 'function';
}

function expectFailure(done, message) {
  return function(err) {
    if (err) done();
    else (done.fail || done)(new Error(message || 'expected failure'));
  };
}

const pending = Object.create(null);
var nextCallbackId = 0;

function dispatch(obj, callback) {
  const id = nextCallbackId++;
  obj = Object.assign({id: id}, obj);
  const frame = document.createElement('iframe');
  pending[id] = {message: obj, frame: frame, callback: callback};
  frame.addEventListener('error', callback);
  frame.src = obj.frameSrc + '#' + id;
  frame.style.display = 'none';
  document.body.appendChild(frame);
}

const hasOwn = Object.prototype.hasOwnProperty;
function receiveMessage(event) {
  const data = event.data;

  if (data.type === 'ready') {
    const meta = pending[data.id];
    if (meta) {
      meta.frame.contentWindow.postMessage(meta.message, '*');
    } else {
      throw new Error('erroneous ready event');
    }
  } else if (data.type === 'callback') {
    const meta = pending[data.id];
    if (meta) {
      const callback = meta.callback;
      const frame = meta.frame;
      delete pending[data.id];
      frame.parentNode.removeChild(frame);
      // console.log('err', data.error);
      if (data.error) (callback.fail || callback)(data.error);
      else callback(null);
    } else {
      throw new Error('erroneous callback');
    }
  }
}

addEventListener('message', receiveMessage, false);


spec/server.js
const bodyParser = require('body-parser');
const express = require('express');
// const request = require('request');
const cors = require('cors');
const url = require('url');
const useragent = require('useragent');
const emptygif = require('emptygif');

const corsGate = require('..');

const alice = express();
const bob = express();
const eve = express();

[alice, bob, eve].forEach((app) => {
  app.set('views', __dirname + '/templates');
  app.set('view engine', 'hbs');

  // Used to verify that `navigator.sendBeacon` can POST JSON even without preflighting.
  app.use(bodyParser.json());

  // Disable caching for all responses.
  app.use(function(req, res, next) {
    // if (req.url !== '/') console.log(req.method, req.query.id, req.url);
    res.setHeader('Cache-Control', 'no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0');
    next();
  });

  // Make sure that we can still use the Origin header with the referrer disabled.
  app.use((req, res, next) => {
    const agent = useragent.parse(req.get('user-agent'));
    if (agent.family === 'Firefox') {
      // allows us to interpret the referer as the origin
      res.set('referrer-policy', 'strict-origin');
    } else if (agent.family === 'Safari') {
      res.locals.metaReferrer = true;
    } else {
      res.set('referrer-policy', 'no-referrer');
    }
    next();
  });

  app.disable('x-powered-by');

  app.get('/', (req, res) => {
    res.render('proxy-page');
  });
});

const lockDown = [
  // We must update the origin correctly here so that the access-control-allow-origin response
  // header correctly reflects whether the request should be allowed.
  corsGate.originFallbackToReferrer(),
  cors({
    origin: ['http://localhost:9997', 'http://localhost:9998'],
    credentials: true
  }),
  corsGate({
    origin: 'http://localhost:9997',
    strict: false,
    allowSafe: true
  })
];

alice.use('/api', lockDown, (req, res) => {
  if (req.path.endsWith('image')) {
    // We send a real image so that the image error handler will reflect whether the server rejected
    // the response.
    emptygif.sendEmptyGif(req, res, {
      'Content-Type': 'image/gif',
      'Content-Length': emptygif.emptyGifBufferLength,
      'Cache-Control': 'no-cache, no-store, must-revalidate, max-age=0' // Never cache.
    });
  } else {
    res.end()
  }
});

var servers = [];

function start(unref) {
  servers.push(
    alice.listen(9997),
    bob.listen(9998),
    eve.listen(9999));

  if (unref) {
    servers.forEach(function(server) {
      server.unref();
    });
  }
}

function stop() {
  servers.forEach(function(server) {
    server.close();
  });
}

exports.start = start;
exports.stop = stop;

if (require.main === module) start();


test/cors-gate.spec.js
const cors = require('cors');
const corsGate = require('..');
const request = require('supertest');
const express = require('express');

function ok(req, res) {
  res.status(200).end();
}

describe('cors-gate', function() {
  beforeEach(function() {
    this.app = express();
  });

  it('should allow same-origin requests', function(done) {
    this.app.use(corsGate({
      allowSafe: false,
      origin: 'http://localhost'
    }));

    this.app.get('/get', ok);

    request(this.app)
      .get('/get')
      .set('origin', 'http://localhost')
      .expect(200, done);
  });

  it('should allow same-origin POST requests', function(done) {
    this.app.use(corsGate({
      allowSafe: false,
      origin: 'http://localhost'
    }));

    this.app.post('/post', ok);

    request(this.app)
      .post('/post')
      .set('origin', 'http://localhost')
      .expect(200, done);
  });

  it('should allow permitted cross-origin requests', function(done) {
    this.app.use(cors({
      origin: 'http://localhost:8080'
    }), corsGate({
      allowSafe: false,
      origin: 'http://localhost'
    }));

    this.app.post('/post', ok);

    request(this.app)
      .post('/post')
      .set('origin', 'http://localhost:8080')
      .expect(200, done);
  });

  it('should allow wildcard origins', function(done) {
    this.app.use(cors({
      origin: '*'
    }), corsGate({
      origin: 'http://localhost'
    }));

    this.app.post('/post', ok);

    request(this.app)
      .post('/post')
      .set('origin', 'http://localhost:8080')
      .expect(200, done);
  });

  it('should terminate on preflight requests', function(done) {
    this.app.use(cors({
      origin: '*'
    }), corsGate({
      origin: 'http://localhost'
    }));

    this.app.post('/post', ok);

    request(this.app)
      .options('/post')
      .set('origin', 'http://localhost:8080')
      .expect('Content-Length', '0')
      .expect(204, done);
  });

  it('should reject requests without origin', function(done) {
    this.app.use(corsGate({
      origin: 'http://localhost'
    }));

    this.app.post('/post', ok);

    request(this.app)
      .post('/post')
      .expect(403, done);
  });

  it('should reject requests from other origins', function(done) {
    this.app.use(cors({
      origin: 'http://localhost:8080'
    }), corsGate({
      origin: 'http://localhost'
    }));

    this.app.post('/post', ok);

    request(this.app)
      .post('/post')
      .set('origin', 'https://mixmax.com')
      .expect(403, done);
  });

  describe('options.allowSafe', function() {
    it('should allow unspecified safe requests', function(done) {
      this.app.use(corsGate({
        strict: true,
        allowSafe: true,
        origin: 'http://localhost'
      }));

      this.app.get('/get', ok);

      request(this.app)
        .get('/get')
        .expect(200, done);
    });

    it('should reject unspecified safe requests', function(done) {
      this.app.use(corsGate({
        strict: true,
        allowSafe: false,
        origin: 'http://localhost'
      }));

      this.app.get('/get', ok);

      request(this.app)
        .get('/get')
        .expect(403, done);
    });

    it('can be given a function returning false', function(done) {
      this.app.use(corsGate({
        strict: true,
        allowSafe: () => false,
        origin: 'http://localhost'
      }));

      this.app.get('/get', ok);

      request(this.app)
        .get('/get')
        .expect(403, done);
    });

    it('can be given a function returning true', function(done) {
      this.app.use(corsGate({
        strict: true,
        allowSafe: () => true,
        origin: 'http://localhost'
      }));

      this.app.get('/get', ok);

      request(this.app)
        .get('/get')
        .expect(200, done);
    });
  });

  describe('options.strict', function() {
    it('should permit requests without origin', function(done) {
      this.app.use(corsGate({
        strict: false,
        origin: 'http://localhost'
      }));

      this.app.post('/post', ok);

      request(this.app)
        .post('/post')
        .expect(200, done);
    });
  });

  describe('options.failure', function() {
    it('should not be invoked for same-origin requests', function(done) {
      this.app.use(corsGate({
        allowSafe: false,
        origin: 'http://localhost',
        failure: function(req, res) {
          res.status(403).end();
        }
      }));

      this.app.post('/post', ok);

      request(this.app)
        .post('/post')
        .set('origin', 'http://localhost')
        .expect(200, done);
    });


    it('should be invoked for requests without origin', function(done) {
      this.app.use(corsGate({
        origin: 'http://localhost',
        failure: function(req, res) {
          res.status(204).end();
        }
      }));

      this.app.post('/post', ok);

      request(this.app)
        .post('/post')
        .expect(204, done);
    });

    it('should be invoked for requests from other origins', function(done) {
      this.app.use(cors({
        origin: 'http://localhost:8080'
      }), corsGate({
        origin: 'http://localhost',
        failure: function(req, res) {
          res.status(204).end();
        }
      }));

      this.app.post('/post', ok);

      request(this.app)
        .post('/post')
        .set('origin', 'https://mixmax.com')
        .expect(204, done);
    });
  });

  describe('originFallbackToReferrer', function() {
    it('should patch origin', function(done) {
      this.app.use(corsGate.originFallbackToReferrer());

      this.app.get('/get', function(req, res) {
        const correctOrigin = req.get('origin') === 'http://localhost';
        res.status(correctOrigin ? 200 : 403).end();
      });

      request(this.app)
        .get('/get')
        .set('referer', 'http://localhost/home')
        .expect(200, done);
    });

    it('should not overwrite origin', function(done) {
      this.app.use(corsGate.originFallbackToReferrer());

      this.app.get('/get', function(req, res) {
        const correctOrigin = req.get('origin') === 'http://localhost';
        res.status(correctOrigin ? 200 : 403).end();
      });

      // This scenario won't technically occur in the wild - the referer and origin should always
      // match.
      request(this.app)
        .get('/get')
        .set('referer', 'http://localhost:8080/home')
        .set('origin', 'http://localhost')
        .expect(200, done);
    });
  });
});


index.js
const url = require('url');

/**
 * Called when we successfully gate a request. If the
 * request is an OPTION request, terminate the request here.
 * Otherwise, pass on to the next middleware.
 */
function success(req, res, next) {
  if (req.method === 'OPTIONS') {
    // If this is an OPTIONS request, terminate here.
    res.statusCode = 204;
    // Safari needs a content-length for 204, see https://github.com/expressjs/cors/blob/master/lib/index.js#L176
    res.setHeader('Content-Length', '0');
    res.end();
  } else {
    next();
  }
}

/**
 * Gate requests based on CORS data. For requests that are not permitted via CORS, invoke the
 * failure options callback, which defaults to rejecting the request.
 *
 * @param {Object} options
 * @param {String} options.origin The origin of the server - requests from this origin will always
 *   proceed.
 * @param {Boolean=} options.strict Whether to reject requests that lack an Origin header. Defaults
 *   to true.
 * @param {Boolean|Function=} options.allowSafe Whether to enforce the strict mode for safe requests (HEAD,
 *   GET). Defaults to true. A function can be passed here, that will be passed req and res and should return a boolean.
 * @param {Function=} options.failure A standard connect-style callback for handling failure.
 *   Defaults to rejecting the request with 403 Unauthorized.
 */
function corsGate(options) {
  options = Object.assign({
    strict: true,
    allowSafe: true,
    failure(req, res, next) {
      // Set `statusCode` vs. using `res#status`, as https://github.com/expressjs/cors does, so this
      // will work with any Connect-compatible server.
      res.statusCode = 403;
      res.end();
    }
  }, options);

  if (typeof options.origin !== 'string') {
    throw new Error("Must specify the server's origin.");
  }

  const thisOrigin = options.origin.toLowerCase();
  const failure = options.failure;

  return function(req, res, next) {
    const origin = (req.headers.origin || '').toLowerCase().trim();

    if (!origin) {
      const allowSafe = !!(typeof options.allowSafe === 'function' ? options.allowSafe(req, res) : options.allowSafe);
      // Fail on missing origin when in strict mode, but allow safe requests if allowSafe set.
      if (options.strict && (!allowSafe || ['GET', 'HEAD'].indexOf(req.method) === -1)) {
        return failure(req, res, next);
      }

      return success(req, res, next);
    }

    // Always allow same-origin requests.
    if (origin === thisOrigin) return success(req, res, next);

    // Now this is a cross-origin request. Check if we should allow it based on headers set by
    // previous CORS middleware. Note that `getHeader` is case-insensitive.
    const otherOrigin = (res.getHeader('access-control-allow-origin') || '').toLowerCase().trim();

    // Two values: allow any origin, or a specific origin.
    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Access-Control-Allow-Origin
    if ((otherOrigin === '*') || (origin === otherOrigin)) return success(req, res, next);

    // CSRF! Abort.
    failure(req, res, next);
  };
}

/**
 * If the Origin header is missing, fill it with the origin part of the Referer.
 *
 * Firefox does not send the Origin header for same-origin requests, as of version 53. This is a
 * documented bug, so this middleware enables verification of the Origin in that case. Additionally,
 * no browser sends the Origin header when sending a GET request to load an image. We could simply
 * allow all GET requests - GET requests are safe, per HTTP - but we'd rather reject unauthorized
 * cross-origin GET requests wholesale.
 *
 * At present, Chrome and Safari do not support the strict-origin Referrer-Policy, so we can only
 * patch the Origin from the Referer on Firefox. In patching it, however, we can reject unauthorized
 * cross-origin GET requests from images, and once Chrome and Safari support strict-origin, we'll
 * be able to do so on all three platforms.
 *
 * In order to actually reject these requests, however, the patched Origin data must be visible to
 * the cors middleware. This middleware is distinct because it must appear before cors and corsGate
 * to perform all the described tasks.
 */
function originFallbackToReferrer() {
  return function(req, res, next) {
    const origin = req.headers.origin;
    if (!origin) {
      const ref = req.headers.referer;
      if (ref) {
        const parts = url.parse(ref);
        req.headers.origin = url.format({
          protocol: parts.protocol,
          host: parts.host
        });
      }
    }
    next();
  };
}

corsGate.originFallbackToReferrer = corsGate.originFallbackToReferer = originFallbackToReferrer;

module.exports = corsGate;
